File: packages/main/src/tools/arch.ts
import { z } from 'zod'
import type { ToolDefinition } from '@buildoto/opencode-core/tool'
import { defineFreecadTool } from './registry'
import { toolInvoke } from '../freecad/client'

const POINT3 = z.tuple([z.number(), z.number(), z.number()])

const wallSchema = z.object({
  length_mm: z.number().positive(),
  height_mm: z.number().positive(),
  thickness_mm: z.number().positive(),
  alignment: z.enum(['left', 'center', 'right']).optional(),
  position: POINT3.optional(),
})

const floorSchema = z.object({
  polygon: z.array(POINT3).min(3),
  thickness_mm: z.number().positive(),
})

const windowSchema = z.object({
  host_wall_id: z.string(),
  width_mm: z.number().positive(),
  height_mm: z.number().positive(),
  sill_height_mm: z.number().nonnegative(),
  offset_along_wall_mm: z.number(),
})

const doorSchema = z.object({
  host_wall_id: z.string(),
  width_mm: z.number().positive(),
  height_mm: z.number().positive(),
  offset_along_wall_mm: z.number(),
})

const roofSchema = z.object({
  base_object_id: z.string(),
  angle_deg: z.number(),
  thickness_mm: z.number().positive(),
})

export const archTools: ToolDefinition[] = [
  defineFreecadTool({
    id: 'arch_create_wall',
    description:
      'Create a parametric BIM wall (length, height, thickness in mm). Alignment: left | center | right (centerline relative to base line).',
    provenance: 'freecad',
    inputSchema: wallSchema,
    async handler(input) {
      const data = await toolInvoke('arch_create_wall', input)
      return JSON.stringify(data)
    },
  }),
  defineFreecadTool({
    id: 'arch_create_floor',
    description:
      'Create a BIM floor / slab from a closed polygon footprint and a thickness in mm.',
    provenance: 'freecad',
    inputSchema: floorSchema,
    async handler(input) {
      const data = await toolInvoke('arch_create_floor', input)
      return JSON.stringify(data)
    },
  }),
  defineFreecadTool({
    id: 'arch_create_window',
    description:
      'Cut a window into a host wall. Offsets are in millimetres along and up the wall.',
    provenance: 'freecad',
    inputSchema: windowSchema,
    async handler(input) {
      const data = await toolInvoke('arch_create_window', input)
      return JSON.stringify(data)
    },
  }),
  defineFreecadTool({
    id: 'arch_create_door',
    description:
      'Cut a door into a host wall. Offset is along the wall in millimetres; door sits at the wall base.',
    provenance: 'freecad',
    inputSchema: doorSchema,
    async handler(input) {
      const data = await toolInvoke('arch_create_door', input)
      return JSON.stringify(data)
    },
  }),
  defineFreecadTool({
    id: 'arch_create_roof',
    description:
      'Create a BIM roof from a base wire / face object id, with slope angle (degrees) and thickness (mm).',
    provenance: 'freecad',
    inputSchema: roofSchema,
    async handler(input) {
      const data = await toolInvoke('arch_create_roof', input)
      return JSON.stringify(data)
    },
  }),
]

File: resources/freecad/handlers/arch.py
"""Arch (BIM) workbench handlers."""

from __future__ import annotations

import FreeCAD  # type: ignore

try:
    import Arch  # type: ignore
except Exception:  # pragma: no cover
    Arch = None  # type: ignore

try:
    import Draft  # type: ignore
except Exception:  # pragma: no cover
    Draft = None  # type: ignore

try:
    import Part  # type: ignore
except Exception:  # pragma: no cover
    Part = None  # type: ignore


_ALIGNMENTS = {"left": "Left", "center": "Center", "right": "Right"}


def _vec(pos, default=(0, 0, 0)):
    v = tuple(pos or default)
    return FreeCAD.Vector(*v)


def _resolve(doc, obj_id):
    """Resolve an object ID, trying direct lookup first, then case-insensitive."""
    obj = doc.getObject(obj_id)
    if obj is not None:
        return obj
    # Case-insensitive fallback — LLMs often lowercase FreeCAD's capitalized Names.
    for o in doc.Objects:
        if o.Name.lower() == obj_id.lower() or o.Label.lower() == obj_id.lower():
            return o
    return None


def _require_arch():
    if Arch is None:
        raise RuntimeError("Arch workbench unavailable in this FreeCAD build")


def arch_create_wall(payload, doc):
    _require_arch()
    length = float(payload["length_mm"])
    height = float(payload["height_mm"])
    thickness = float(payload["thickness_mm"])
    alignment = _ALIGNMENTS.get(payload.get("alignment") or "center", "Center")
    pos = _vec(payload.get("position"))

    if Draft is None:
        raise RuntimeError("Draft workbench required for Arch.makeWall baseline")
    base = Draft.makeLine(pos, pos.add(FreeCAD.Vector(length, 0, 0)))
    wall = Arch.makeWall(base, length=length, width=thickness, height=height)
    wall.Align = alignment
    doc.recompute()
    return {
        "ok": True,
        "object_id": wall.Name,
        "label": wall.Label,
        "length_mm": length,
        "height_mm": height,
        "thickness_mm": thickness,
        "alignment": payload.get("alignment") or "center",
    }


def arch_create_floor(payload, doc):
    _require_arch()
    thickness = float(payload["thickness_mm"])
    polygon = payload.get("polygon")
    if not polygon or len(polygon) < 3:
        raise ValueError("polygon must have at least 3 vertices")
    if Draft is None:
        raise RuntimeError("Draft workbench required for floor baseline")
    points = [FreeCAD.Vector(*p) for p in polygon]
    points.append(points[0])
    wire = Draft.makeWire(points, closed=True, face=True)
    floor = Arch.makeStructure(wire, height=thickness)
    floor.IfcType = "Slab"
    floor.Label = "Floor"
    doc.recompute()
    return {
        "ok": True,
        "object_id": floor.Name,
        "label": floor.Label,
        "thickness_mm": thickness,
    }


def arch_create_window(payload, doc):
    _require_arch()
    host_id = payload["host_wall_id"]
    width = float(payload["width_mm"])
    height = float(payload["height_mm"])
    sill = float(payload["sill_height_mm"])
    offset = float(payload["offset_along_wall_mm"])
    host = _resolve(doc, host_id)
    if host is None:
        raise LookupError(f"unknown host wall: {host_id}")
    window = Arch.makeWindow(host, width=width, height=height)
    try:
        window.HoleDepth = 500
    except Exception:
        pass
    placement = FreeCAD.Placement()
    placement.Base = FreeCAD.Vector(offset, 0, sill)
    window.Placement = placement
    doc.recompute()
    return {
        "ok": True,
        "object_id": window.Name,
        "label": window.Label,
        "host_wall_id": host_id,
    }


def arch_create_door(payload, doc):
    _require_arch()
    host_id = payload["host_wall_id"]
    width = float(payload["width_mm"])
    height = float(payload["height_mm"])
    offset = float(payload["offset_along_wall_mm"])
    host = _resolve(doc, host_id)
    if host is None:
        raise LookupError(f"unknown host wall: {host_id}")
    door = Arch.makeWindow(host, width=width, height=height)
    try:
        door.WindowParts = ["Door", "Frame", "Wire0", "30", "40"]
    except Exception:
        pass
    door.Label = "Door"
    placement = FreeCAD.Placement()
    placement.Base = FreeCAD.Vector(offset, 0, 0)
    door.Placement = placement
    doc.recompute()
    return {
        "ok": True,
        "object_id": door.Name,
        "label": door.Label,
        "host_wall_id": host_id,
    }


def arch_create_roof(payload, doc):
    _require_arch()
    base_id = payload["base_object_id"]
    angle = float(payload["angle_deg"])
    thickness = float(payload["thickness_mm"])
    base = _resolve(doc, base_id)
    if base is None:
        raise LookupError(f"unknown base: {base_id}")
    roof = Arch.makeRoof(base, angle=angle, thickness=thickness)
    doc.recompute()
    return {
        "ok": True,
        "object_id": roof.Name,
        "label": roof.Label,
        "angle_deg": angle,
        "thickness_mm": thickness,
    }


TOOLS = {
    "arch_create_wall": arch_create_wall,
    "arch_create_floor": arch_create_floor,
    "arch_create_window": arch_create_window,
    "arch_create_door": arch_create_door,
    "arch_create_roof": arch_create_roof,
}
